### DISCUSSION TITLE
Piplines Install & Run (Guide)

### DISCUSSION THREAD

**Discussion by rootcli** [open-webui/pipelines#62] [General]
I created this little guide to help newbies Run pipelines, as it was a challenge for me to install and run pipelines. Hope it helps.

##### Quick Start with Docker
[Pipelines Repository Quick Guide](https://github.com/open-webui/pipelines/tree/main#-quick-start-with-docker)

## What to do Next?

1. *Run the Pipelines container:*
    
2. *Login to Open WebUI:*
    - In Open WebUI: Navigate to the **Admin Panel > Settings > Connections > OpenAI API** section.
    - Set the API URL to `http://host.docker.internal:9099` and the API key to `0p3n-w3bu!`. Your pipelines should now be active. IF localhost doesn't work use `host.docker.internal`

 3. *You can Download the examples folder (or individual files, as you wish) from the Pipelines repository:*
 
 `To Download the Examples Folder:
```bash
git clone https://github.com/open-webui/pipelines.git examples
```

`Here we will perform an example on one file which is called wikipedia_pipeline.py you can direct download it to linux from the link below`
```shell
curl -O https://raw.githubusercontent.com/open-webui/pipelines/c9766d630ab69549b06e994b39fb173a61bb82e8/examples/pipelines/integrations/wikipedia_pipeline.py
```

4. Place The downloaded file `wikipedia_pipeline.py` in **pipelines Root Folder.**

2. *In the open webui interface Go to Admin Panel > Gear Icon in top right ⚙ > pipelines*

	you should see `Pipelines Valves` click on the pulldown menu and select `wikipedia_pipeline` 
<img width="861" alt="Pasted image 20240605090833" src="https://github.com/open-webui/pipelines/assets/94934393/3d18818a-3a9d-49e7-8d7f-8d5ed55a864a">

3. Go to New Chat > From the Models pulldown Menu select ``
<img width="968" alt="Pasted image 20240605091159" src="https://github.com/open-webui/pipelines/assets/94934393/c36de630-6e38-4e73-baba-479077d6dbef">

4. Now type a topic you wish to search from `wikipedia

**Comment by santosh0787** [comment]
I really like this feature. Can you please add some documentation on how to add tools functions? I added examples/filters/function_calling_filter_pipeline.py but it is added as filters. Can you tell how invoke these functions? 
All I get is query is going to Wiki pipeline.

**Comment by c9482** [comment]
When I only select "Wikipedia Pipeline", I get nothing. I have to add another model to the chat to get a response and the response does not come from "Wikipedia Pipeline". I was expecting to get the latest entries from wikipedia. Perhaps I am still missing how this is supposed to work? See below.

![image](https://github.com/open-webui/pipelines/assets/17013052/8d93efda-e566-4d43-855d-0d54f0e2aaf7)

**Reply by rootcli** [reply]
"no information found" means the piplines model didn't find information.
try "Sun" or "Moon" in the prompt

**Reply by c9482** [reply]
Thanks, I get the point.

**Reply by hblink** [reply]
and what is the point? i have the same problem!

**Reply by stevenfullman** [reply]
> and what is the point? i have the same problem!

I think the point is that rather than write a "prompt" like "who won the world cup in 2022", the prompt is simply "world cup 2022". That's the only way I can get answers anyway

**Comment by delfireinoso** [comment]
> Place The downloaded file wikipedia_pipeline.py in pipelines Root Folder.

I have the file wikipedia_pipeline.py 
Where I find the Piplelines Root Folder?

**Reply by c9482** [reply]
It's easiest to install to install from Github URL. For manual installation, there's an empty "pipelines" folder in the repo. Or if you're using docker, mount a volume for /app/pipelines.

![image](https://github.com/open-webui/pipelines/assets/17013052/1dd7a852-bd9b-408b-9d58-24f80cfc9dd1)

**Reply by delfireinoso** [reply]
I use Docker as it's easy on a Mac 

To mount a volume I think that I need to modify the Docker install command for Pipelines 

This should come on the install documentation

**Reply by delfireinoso** [reply]
I see that the Docker default install has a Pipelines Volume, I'll tried the url approch and it's fine and working

**Comment by c9482** [comment]
Still couple of unanswered questions about function calling filter pipeline:

1. How does it get activated? For example, function_calling_filter_pipeline example has get_current_time tool. I can't get it to fire. Is there a specific prompt that I need to use?
2. Are the changes to pipeline automatically detected and the pipeline is reloaded? Or do I have to restart the pipelines server?

Many thanks in advance for some clues.

**Reply by atgehrhardt** [reply]
I am running both openwebui and pipelines in docker and for me it's automatically updated if u upload the pipeline via the UI.

Also to get it to fire your request simply needs to reference something related to the function.

If I ask what time it is, it fires the function.

Here's the catch, the language model has to be capable of function calling. 

To my knowledge the only base (not fine tuned) open source model that can do that is Llama3 (though I could be wrong).

If you're using Ollama the functions will NOT work unless you add the url to the pipeline and set the api key as empty

So for me the openai api base url is: http://host.docker.internal:11434/v1

Which accesses ollama on my host system.

**Reply by c9482** [reply]
This is great information, thanks! One other question: how are package dependencies resolved? It seems that only the Python file gets deployed. I am not a Python programmer so this is a bit of a mystery to me.

**Reply by atgehrhardt** [reply]
So I haven't built anything with external dependencies yet but I can test later. Looking through the pipelines code I'm not seeing a way to install dependencies or an automatic mechanism, so may not be implemented yet. You could dial into the docker and install them, but I think that may be problematic as conflicting packages would be an issue since it looks like pipelines aren't running in venvs.

I can't be sure, may be worth opening a discussion on the Pipelines github

**Reply by atgehrhardt** [reply]
Looks like this is addressed here: https://github.com/open-webui/pipelines/discussions/65#discussioncomment-9682144

**Reply by justinh-rahb** [reply]
@atgehrhardt you can add a header to a pipeline such as this, that specifies the dependencies to install:
```python
"""
title: MLX Pipeline
author: justinh-rahb
date: 2024-05-27
version: 1.1
license: MIT
description: A pipeline for generating text using Apple MLX Framework.
requirements: requests, mlx-lm, huggingface-hub
environment_variables: MLX_HOST, MLX_PORT, MLX_MODEL, MLX_STOP, MLX_SUBPROCESS, HUGGINGFACE_TOKEN
"""

from typing import List, Union, Generator, Iterator
from schemas import OpenAIChatMessage
import requests
import os
import subprocess
import logging
from huggingface_hub import login

...
```

**Reply by atgehrhardt** [reply]
Oh that's fantastic, thank you for the clarification!

**Comment by musicheardworldwide** [comment]
If you create tools or functions you want to be triggered, you have to create the filter as well. This took me until today to figure out, but your pipe is in the data, the filter is how it gets to where its going and how. Keep building up your tool pipelines or just tools, and make sure you have the python function filter up and accurate, it should just trigger whenever you say the right thing. also, dont forget every tool you make needs to have a call, so once you get that locked in it should be smooth...

i just had my model message me on icloud via tools. There's a LOT of potential with this

**Reply by santosh0787** [reply]
Can you provide a simple example of how to use pipelines and filters? I've looked at the examples in the repository, but they're scattered among other examples and I'm having trouble understanding how to apply them. Could you provide a clear and concise example to help me get started?

**Reply by nomppy** [reply]
Seconding this. There is function calling blueprint, function calling filter, and a function calling scaffold?

Am I correct in understanding that the difference between a filter and pipeline is that a filter is a middleware inserted between user input and sending it directly to an LLM, while a pipeline creates a "model" that appears in the dropdown?

I see this snippet a lot in the examples, but what does it mean to be connected to a pipeline? My current usage is using the Azure pipeline as the model, so it would be great if I can configure filters to use that pipeline as well. Any clarification would be greatly appreciated!

```
# List target pipeline ids (models) that this filter will be connected to.
# If you want to connect this filter to all pipelines, you can set pipelines to ["*"]
pipelines: List[str] = []
```

**Reply by justinh-rahb** [reply]
> Am I correct in understanding that the difference between a filter and pipeline is that a filter is a middleware inserted between user input and sending it directly to an LLM, while a pipeline creates a "model" that appears in the dropdown?

Yes this is correct.

> I see this snippet a lot in the examples, but what does it mean to be connected to a pipeline? My current usage is using the Azure pipeline as the model, so it would be great if I can configure filters to use that pipeline as well. Any clarification would be greatly appreciated!
> 
> ```
> # List target pipeline ids (models) that this filter will be connected to.
> # If you want to connect this filter to all pipelines, you can set pipelines to ["*"]
> pipelines: List[str] = []
> ```

This is how you determine which "models" or pipelines are acted upon by the filter. By default it will be all (`*`), or you can set model or pipeline IDs, comma-separated.

**Reply by nomppy** [reply]
Thanks very much for the clarification. So for example, if I want to implement function calling for Azure models, I would write the Azure pipeline id in the pipelines attribute for the function calling filter (or leave it as `*`), which will insert the function calling filter between the user's input and the Azure pipeline. 

If I want the _filter itself_ to use Azure OpenAI, e.g. to retrieve the function call parameters, is there a way to have a filter call a pipeline? This might be somewhat of a niche use. The other option I see is to just modify the filter directly to take in Azure keys instead. 

Btw, what are blueprints and manifolds?

**Reply by justinh-rahb** [reply]
Blueprints are just part of the framework's boilerplate code, you shouldn't have any need to mess with these. Manifold is just a special pipeline class that provides more than one pipe back to the models array. So for example an LLM provider manifold will have a pipe for each model it provides.

**Reply by nomppy** [reply]
Given that the current function calling blueprint uses OpenAI as a provider, wouldn't I need to modify it if I want to use a different folder?

**Reply by TheDarkTrumpet** [reply]
> Blueprints are just part of the framework's boilerplate code, you shouldn't have any need to mess with these. Manifold is just a special pipeline class that provides more than one pipe back to the models array. So for example an LLM provider manifold will have a pipe for each model it provides.

This is really helpful to have it explained.  Was messing with pipelines and making raw OpenAI calls trying to figure out what's what in the repo.  There's a good amount of documentation, but I almost feel like the Pipelines project should be treated more separately with their own set of documentation.  That way it's more clear about things that specifically apply here.  The Manifold really threw me for a loop.  I thought it was a way for a way to compartmentalize the  pipelines, and the manifold could call those pipelines as a collection.

**Comment by eladrave** [comment]
I installed the https://github.com/open-webui/pipelines/blob/main/examples/filters/function_calling_filter_pipeline.py

However, I can never get the weather part work.
I tried "weather in new york" , but it didn't use it.

I tested my access to openweathermap, and it worked fine:
curl -G "http://api.openweathermap.org/data/2.5/weather" \
    --data-urlencode "q=New York" \
    --data-urlencode "appid=MY_KEY" \
    --data-urlencode "units=imperial"

**Reply by saccodd** [reply]
does the current time work for you? for me none of them is working. I'm trying to use models provided through ollama (e.g. [llama3.1:latest](https://ollama.com/library/llama3.1) ) instead of the OpenAI endpoint and replacing it with http://host.docker.internal:11434/v1 or http://localhost:11434/v1 but nothing works

**Comment by PlebeiusGaragicus** [comment]
I cannot wait until more documentation and examples come out!!!

Right now I'm thinking I need a filter to track user token usage.  When the user has used all their allotted tokens I need the AI model to respond with a payment request LNURL.  I made my own webui scratch to do this, but now that Open WebUI includes this AWESOME custom extension ability I'm going to pivot towards this.

This could be THE feature I've been looking for in the ecosystem to build this out.  Self-hosted, micropayment-enabled, custom LangGraph AI constructs for the rest of us...

**Comment by VfBfoerst** [comment]
If you want to run and connect the pipelines server with **Podman** , you need to slightly adjust the "run" command: `host.docker.internal` to `host.containers.internal`:  
```bash
podman run --name pipelines -d -p 9099:9099 --add-host=host.containers.internal:host-gateway -v /path/on/host/pipelines:/app/pipelines --restart always ghcr.io/open-webui/pipelines:main
```
In the admin-Settings, you will need to do the same (`http://host.containers.internal:9099`):   
![image](https://github.com/user-attachments/assets/4ae8b62d-564e-46db-b6a2-fd3d856e61a8)

**Reply by alifarooq93** [reply]
hi, can i add the pipelines connection http://host.containers.internal:9099 into ollama connection? I tried it but its not verifying the connection but if i add into open ai it successfully verified. 
I am not using openai so after adding pipelines they are not working saying connection closed. I checked the docker open webUI container and it said missing open ai key therefore I need to connect pipelines with ollama. Is it possible?

**Comment by saccodd** [comment]
Trying to run https://github.com/open-webui/pipelines/blob/main/examples/filters/function_calling_filter_pipeline.py but still unsuccessful.

I am using Open WebUI with Bundled Ollama Support (see https://github.com/open-webui/open-webui?tab=readme-ov-file#installing-open-webui-with-bundled-ollama-support). Checking Docker logs I realised that Open WebUi was not able to connect to Ollama API (even if the downloaded models were working fine). So I reached the following setup that doesn't throw errors.

Launch container with:
`docker run -d -p 4000:8080 --add-host=host.docker.internal:host-gateway -v ollama:/root/.ollama -v open-webui:/app/backend/data -e OLLAMA_BASE_URL=http://host.docker.internal:11434 --name open-webui --restart always ghcr.io/open-webui/open-webui:ollama`

Setup API endpoints as follows:
![image](https://github.com/user-attachments/assets/38d59559-0097-4b88-a751-cbeba393ef4d)

I installed the pipeline from Github URL and setup as follows:
![image](https://github.com/user-attachments/assets/da9b9c60-44c6-4473-876b-f422b331fb8e)

If I use localhost instead of host.docker.internal as OpenAI API Base URL, still no error is thrown, but I can't get my models to use the pipeline.
Currently testing with [llama3.1:latest](https://ollama.com/library/llama3.1) and [llama3-groq-tool-use:latest](https://ollama.com/library/llama3-groq-tool-use)

**Reply by nomppy** [reply]
So the issue is what exactly? That you can't connect to Ollama or that the function calling filter isn't being triggered?

**Reply by saccodd** [reply]
function calling filter pipeline is not being triggered

**Reply by Urvesh71** [reply]
I also have executed Open Web UI in a docker container along with Ollama container running parallel. Apart from that I have executed the Pipeline container as well. And I followed the steps of adding http://host.docker.internal:9099 and 0p3n-w3bu! as API key. I varified successfully. But still I cannot see pipeline. In my Pipeline section, It still shows 'Pipelines Not Detected'.

**Comment by mythosofllama** [comment]
Hello,
I am currently trying to integrate Ollama into OpenWebUi with a pipeline. 

My model also shows a connection to the pipeline, but gives the following error (when using Wikipedia_Pipeline.py from the GIT-pipelines-examples):
![Ollama_Wikipedia-Pipeline](https://github.com/user-attachments/assets/e1e7efe7-271b-4590-9d99-9eb0dd640865)

Ich benutze Ollama3.2:latest.

The API connections are as follows:
![image](https://github.com/user-attachments/assets/0a6b8fe6-6a17-48e8-902d-41731d48d448)

All connections to the servers can be established.

I run it on wsl with docker.

How can I fix this?

**Reply by Urvesh71** [reply]
Did you figure out any solution to fix your problem? I am also having the same problem but in other pipeline: text_to_sql_pipeline.py

**Reply by Urvesh71** [reply]
@mythosofllama check logs of pipeline container. I tried to implement the same and it works on my side

**Comment by F041** [comment]
I installed the wikipedia pipeline and the [text to sql](https://github.com/open-webui/pipelines/blob/main/examples/pipelines/rag/text_to_sql_pipeline.py). I noticed I didn't see the latter in valves, so I repasted the URL for the installation multiple times. Result: it completed bricked OpenWebUI and in the docker logs I see
```
  File "/usr/local/lib/python3.11/site-packages/fastapi/routing.py", line 278, in app

    raw_response = await run_endpoint_function(

                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.11/site-packages/fastapi/routing.py", line 191, in run_endpoint_function

    return await dependant.call(**values)

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/app/backend/open_webui/main.py", line 1073, in get_models

    models = await get_all_models()

             ^^^^^^^^^^^^^^^^^^^^^^

  File "/app/backend/open_webui/main.py", line 929, in get_all_models

    ollama_models = await get_ollama_models()

                    ^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/app/backend/open_webui/apps/ollama/main.py", line 219, in get_all_models

    "models": merge_models_lists(

              ^^^^^^^^^^^^^^^^^^^

  File "/app/backend/open_webui/apps/ollama/main.py", line 196, in merge_models_lists

    for idx, model_list in enumerate(model_lists):

  File "/app/backend/open_webui/apps/ollama/main.py", line 221, in <lambda>

    lambda response: response["models"] if response else None, responses
```
And I have no idea on what to do

**Reply by Urvesh71** [reply]
Would you please provide the steps you followed ? I would help you but cant figure out what is the problem and what you have done before

**Reply by F041** [reply]
I pasted the RAW URL of the github of text_to_SQL in the Pipeline part of OWUI. Did it a few times for the reason wrote, refreshed the site, then broke. After what I wrote I did a 
```
docker-compose down --volumes
docker-compose build --no-cache
docker-compose up -d
```
and now I get 
```
INFO  [open_webui.apps.ollama.main] get_all_models()

ERROR [open_webui.apps.ollama.main] Connection error: 

ERROR [open_webui.apps.ollama.main] Connection error: 
```

**Reply by Urvesh71** [reply]
Regarding the text_to_sql pipelines: Did you already set the Postgres or SQL database somewhere ? If yes, then in the text_to_sql_pipeline.py, did you change the host details of Database?
This might be one of the reason.

**Reply by F041** [reply]
no I didn't even touch the ENV

**Reply by Urvesh71** [reply]
for text_to_sql_pipeline, you must first set the database locally or on cloud then change the ENV according to your database properties in text_to_sql_pipeline.py. Then it should work.

**Reply by F041** [reply]
but now what can I do to restore it before the bricking point?

**Reply by Urvesh71** [reply]
Delete the Image, Container and its volume and then Run the Docker compose file to rebuild the container

**Reply by F041** [reply]
the volume of just the pipeline? Or of all 3 services? (OWUI, Ollama, Pipelines)

**Reply by Urvesh71** [reply]
delete the Volume of entire container. Start with fresh container

**Reply by F041** [reply]
thanks. Do you have an easy backup solution so it won't happen again? I gave a look at FreeFileSync, Duplicati. I also saw other two docker based solutions, but I prefer GUI

**Comment by CraftzAdmin** [comment]
Hi everyone! I have not found an answer for my problem in any other trackers. I have installed OpenwebUI on a Docker and Ollama directly on a Debian 12 server, and all are working fine. 
I've added the Pipelines with the install script. The container is good, and the connection to the APIs all work. The problem is, when I try to Install from the Github URL, I get a message: 

_**Directory ./data does not exist.**_

Any suggestions? 

![2024-12-12_05h19_15](https://github.com/user-attachments/assets/be014afb-076f-4ea5-8c9a-2302a97fa881)

**Reply by tunganhem2018** [reply]
create directory ./data at your directory which docker-compose.yml file located. You may need to set .env variable like this

DATA_DIR={YOUR-DOCKER-COMPOSE.YML FILE LOCATED}/data
UPLOAD_DIR={YOUR-DOCKER-COMPOSE.YML FILE LOCATED}/data/uploads
CACHE_DIR={YOUR-DOCKER-COMPOSE.YML FILE LOCATED}/data/cache

**Comment by Bobyyy70** [comment]
Hey,

Thing is i deploy it via coolify

Le dim. 12 janv. 2025 à 07:53, tunganhem2018 ***@***.***> a
écrit :
